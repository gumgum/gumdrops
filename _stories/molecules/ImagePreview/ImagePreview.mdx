import { Story, Props } from '@storybook/addon-docs/blocks';
import ImagePreview from '../../../components/molecules/ImagePreview';

# ImagePreview

The `<ImagePreview>` component is used to preview images or gifs.

It will pre-load the image and show a loading indicator until resolved, then the preview will display.


## ImagePreview example
```jsx
<ImagePreview src="PATH_TO_IMAGE">
  <Icon icon="plus" />
</ImagePreview>
```

## Changing the height of the preview
```jsx
<ImagePreview src="PATH_TO_IMAGE" maxHeight={100}>
  <Icon icon="plus" />
</ImagePreview>
```

## Component Example
<Story
    id="molecules-imagepreview--default"
/>

## Props
<Props of={ImagePreview} />
