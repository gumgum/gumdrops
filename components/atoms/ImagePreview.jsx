import React, { useRef, useEffect, useState, useCallback } from 'react';
import ReactDOM from 'react-dom';
import PropTypes from 'prop-types';
import cx from 'classnames';

const LOADING_DIMENSIONS = {
    width: 50,
    height: 20
};

const FAILED_DIMENSIONS = {
    width: 150,
    height: 25
};
function ImagePreview({
    children,
    src,
    alt,
    maxHeight,
    padding,
    position = 'top',
    showArrow = true,
    arrowSize = 5,
    style: customStyles = {},
    tooltipStyles: customTooltipStyles = {},
    className,
    ...otherProps
}) {
    const triggerRef = useRef();
    const [isLoaded, setIsLoaded] = useState(false);
    const [isHovering, setIsHovering] = useState(false);
    const [width, setWidth] = useState(0);
    const [height, setHeight] = useState(0);
    const [styles, setStyles] = useState({});
    const [hasFailed, setHasFailed] = useState(false);

    const baseClass = 'gds-image-preview';

    const rootClass = cx(baseClass, className, {
        [`${baseClass}--loaded`]: isLoaded,
        [`${baseClass}--failed`]: hasFailed,
        [`${baseClass}--show-arrow`]: showArrow,
    });

    useEffect(
        () => {
            const img = new Image();
            setIsLoaded(false);
            setHasFailed(false);
            img.onload = () => {
                setIsLoaded(true);
                setHasFailed(false);
                setWidth(img.width);
                setHeight(img.height);
            };
            img.onerror = () => {
                setIsLoaded(true);
                setHasFailed(true);
                setWidth(FAILED_DIMENSIONS.width);
                setHeight(FAILED_DIMENSIONS.height);
            };
            img.src = src;
        },
        [src]
    );
    // calculate the width based on the height to match the original aspect ratio
    const newWidth = isLoaded ? maxHeight / height * width : 0;
    // simple listener to set a flag when we're hovering
    const onEnterLeave = useCallback(isHovering => setIsHovering(isHovering), []);

    const getTooltipPosition = useCallback(
        () => {
            let tooltipX, tooltipY;
            const el = triggerRef.current;
            const tooltipWidth = hasFailed ? width : isLoaded ? newWidth : LOADING_DIMENSIONS.width;
            const tooltipHeight = hasFailed ? height : isLoaded ? maxHeight : LOADING_DIMENSIONS.height;
            const {
                top,
                left,
                width: _width,
                right,
                height: _height
            } = el.getBoundingClientRect();
            const arrowOffset = showArrow ? arrowSize * 2 : 0;
            const hoverTop = top + window.scrollY;
            const offset = !isLoaded || hasFailed ? 0 : padding;

            switch (position) {
                case 'bottom':
                    tooltipX =
                        left -
                        offset +
                        _width / 2 -
                        tooltipWidth / 2;
                    tooltipY = hoverTop + _height + arrowOffset;
                    break;
                case 'left':
                    tooltipX = left - tooltipWidth - (offset * 2) - arrowOffset;
                    tooltipY =
                        hoverTop + _height / 2 - tooltipHeight / 2 - offset;
                    break;
                case 'right':
                    tooltipX = right + arrowOffset;
                    tooltipY =
                        hoverTop + _height / 2 - tooltipHeight / 2 - offset;
                    break;
                case 'top':
                default:
                    tooltipX =
                        left -
                        offset +
                        _width / 2 -
                        tooltipWidth / 2;
                    tooltipY = hoverTop - tooltipHeight - offset * 2 - arrowOffset;
                    break;
            }
            if (
                isFinite(tooltipY) &&
                isFinite(tooltipX)
            ) {
                return {
                    left: tooltipX,
                    top: tooltipY,
                    width: isLoaded && !hasFailed ? 'auto' : `${tooltipWidth}px`,
                    height: isLoaded && !hasFailed ? 'auto' : `${tooltipHeight}px`,
                };
            }
            return {};
        },
        [
            maxHeight,
            newWidth,
            width,
            height,
            padding,
            isLoaded,
            hasFailed,
            position,
            arrowSize,
            showArrow
        ]
    );

    useEffect(
        () => {
            if (triggerRef && triggerRef.current) {
                setStyles(getTooltipPosition());
            }
        },
        [triggerRef, getTooltipPosition, isHovering]
    );

    const getOppositeSide = useCallback(
        () => {
            switch (position) {
                case 'bottom':
                    return 'top';
                case 'left':
                    return 'right';
                case 'right':
                    return 'left';
                case 'top':
                default:
                    return 'bottom';
            }
        },
        [position]
    );

    const getArrowStyles = useCallback(
        (position, tooltipStyles) => {
            switch (position) {
                case 'bottom':
                    return {
                        bottom: '100%',
                        left: '50%',
                        transform: 'translateX(-50%)',
                        borderLeft: `${arrowSize}px solid transparent`,
                        borderRight: `${arrowSize}px solid transparent`,
                        borderBottom: `${arrowSize}px solid ${tooltipStyles.backgroundColor}`
                    };
                case 'right':
                    return {
                        top: '50%',
                        right: '100%',
                        transform: 'translateY(-50%)',
                        borderTop: `${arrowSize}px solid transparent`,
                        borderBottom: `${arrowSize}px solid transparent`,
                        borderRight: `${arrowSize}px solid ${tooltipStyles.backgroundColor}`
                    };
                case 'left':
                    return {
                        top: '50%',
                        left: '100%',
                        transform: 'translateY(-50%)',
                        borderTop: `${arrowSize}px solid transparent`,
                        borderBottom: `${arrowSize}px solid transparent`,
                        borderLeft: `${arrowSize}px solid ${tooltipStyles.backgroundColor}`
                    };
                case 'top':
                default:
                    return {
                        top: '100%',
                        left: '50%',
                        transform: 'translateX(-50%)',
                        borderLeft: `${arrowSize}px solid transparent`,
                        borderRight: `${arrowSize}px solid transparent`,
                        borderTop: `${arrowSize}px solid ${tooltipStyles.backgroundColor}`
                    };
            }
        },
        [position, arrowSize]
    );

    const tooltipElement = () => {
        const tooltipStyles = {
            position: 'absolute',
            zIndex: 9999,
            backgroundColor: 'rgba(0, 0, 0, 0.8)',
            color: '#fff',
            padding: `${padding}px`,
            borderRadius: 3,
            transition: '0.4s ease',
            transitionProperty: 'transform, opacity, padding',
            transformOrigin: `${getOppositeSide(position)} center`,
            transform: 'scale(0)',
            opacity: 0,
            ...customTooltipStyles,
        }
        return (
            <div
                data-testid="image-preview-tooltip"
                className="gds-flex gds-flex--content-center gds-flex--justify-center"
                style={tooltipStyles}>
                {isLoaded && !hasFailed ? (
                    <img
                        data-testid="image-preview-tooltip-image"
                        src={src}
                        alt={alt}
                        style={{ width: `${newWidth}px`, height: `${maxHeight}px`, display: 'block' }}
                    />
                ) : hasFailed ? null : (
                    <div className="gds-loading__dot gds-loading__dot--sm gds-loading__dot--white" style={{
                        position: 'relative',
                        top: 'auto',
                        left: 'auto',
                    }} />
                )}
                {hasFailed && (
                    <p
                        data-testid="image-preview-load-failure"
                        className="-text-center gds-text--body-xs">
                        Failed to load image.
                    </p>
                )}
                {showArrow && (
                    <span
                        data-testid="image-preview-arrow"
                        style={{
                            position: 'absolute',
                            width: 0,
                            height: 0,
                            ...getArrowStyles(position, tooltipStyles)
                        }}
                    />
                )}
            </div>
        );
    };
    return (
        <span
            data-testid="image-preview-hover-element"
            onMouseEnter={() => onEnterLeave(true)}
            onMouseLeave={() => onEnterLeave(false)}
            ref={triggerRef}
            style={{
                position: 'relative',
                cursor: 'pointer',
                display: 'inline-block',
                ...customStyles
            }}
            className={rootClass}
            {...otherProps}>
            {children}
            {ReactDOM.createPortal(
                React.cloneElement(tooltipElement(), {
                    style: {
                        ...tooltipElement().props.style,
                        ...styles,
                        transform: isHovering ? 'scale(1)' : 'scale(0)',
                        opacity: isHovering ? 1 : 0
                    }
                }),
                document.body
            )}
        </span>
    );
}
ImagePreview.displayName = 'ImagePreview';
ImagePreview.defaultProps = {
    maxHeight: 200,
    padding: 5,
    position: 'top',
    showArrow: true,
    arrowSize: 5
};
ImagePreview.propTypes = {
    children: PropTypes.node.isRequired,
    /** the url to the image */
    src: PropTypes.string.isRequired,
    /** the alt text for the image */
    alt: PropTypes.string,
    /** the maximum height of the image, width is calculated automatically */
    maxHeight: PropTypes.number,
    /** the padding around the image in the tooltip */
    padding: PropTypes.number,
    /** the position of the tooltip relative to the element */
    position: PropTypes.oneOf(['top', 'bottom', 'left', 'right']),
    /** whether or not to show the arrow */
    showArrow: PropTypes.bool,
    /** the size of the arrow in pixels */
    arrowSize: PropTypes.number,
    /** custom styles for the wrapper */
    style: PropTypes.object,
    /** any additional classnames */
    className: PropTypes.string,
    /** custom styles for the tooltip */
    tooltipStyles: PropTypes.object,
};

export default ImagePreview;
