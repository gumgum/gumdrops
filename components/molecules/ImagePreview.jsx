import React, { useRef, useEffect, useState, useCallback } from 'react';
import ReactDOM from 'react-dom';
import PropTypes from 'prop-types';
import LoadingDots from '../atoms/LoadingDots';

const LOADING_DIMENSIONS = {
    width: 50,
    height: 10
};
function ImagePreview({
    children,
    src,
    alt,
    maxHeight,
    padding,
    position = 'top',
    showArrow = true,
    arrowSize = 5,
    style: customStyles = {}
}) {
    const triggerRef = useRef();
    const [isLoaded, setIsLoaded] = useState(false);
    const [isHovering, setIsHovering] = useState(false);
    const [width, setWidth] = useState(0);
    const [height, setHeight] = useState(0);
    const [styles, setStyles] = useState({});
    const [hasFailed, setHasFailed] = useState(false);

    useEffect(
        () => {
            const img = new Image();
            img.onload = () => {
                setIsLoaded(true);
                setHasFailed(false);
                setWidth(img.width);
                setHeight(img.height);
            };
            img.onerror = () => {
                setIsLoaded(true);
                setHasFailed(true);
                setWidth(150);
                setHeight(20);
            };
            img.src = src;
        },
        [src]
    );

    const newWidth = isLoaded ? maxHeight / height * width : 0;

    const onEnterLeave = useCallback(isHovering => setIsHovering(isHovering), []);

    const getTooltipPosition = useCallback(
        () => {
            const el = triggerRef.current;
            const hoveredElementRect = el.getBoundingClientRect();
            const tooltipWidth = hasFailed ? width : isLoaded ? newWidth : LOADING_DIMENSIONS.width;
            const tooltipHeight = hasFailed
                ? height
                : isLoaded ? maxHeight : LOADING_DIMENSIONS.height;
            let tooltipX, tooltipY;
            const arrowOffset = showArrow ? arrowSize * 2 : 0;
            let hoverTop = el.offsetTop + (el.offsetParent ? el.offsetParent.offsetTop : 0);

            switch (position) {
                case 'bottom':
                    tooltipX =
                        hoveredElementRect.left -
                        padding +
                        hoveredElementRect.width / 2 -
                        tooltipWidth / 2;
                    tooltipY = hoverTop + hoveredElementRect.height + arrowOffset;
                    break;
                case 'left':
                    tooltipX = hoveredElementRect.left - tooltipWidth - padding * 2 - arrowOffset;
                    tooltipY =
                        hoverTop + hoveredElementRect.height / 2 - tooltipHeight / 2 - padding;
                    break;
                case 'right':
                    tooltipX = hoveredElementRect.right + arrowOffset;
                    tooltipY =
                        hoverTop + hoveredElementRect.height / 2 - tooltipHeight / 2 - padding;
                    break;
                case 'top':
                default:
                    tooltipX =
                        hoveredElementRect.left -
                        padding +
                        hoveredElementRect.width / 2 -
                        tooltipWidth / 2;
                    tooltipY = hoverTop - tooltipHeight - padding * 2 - arrowOffset;
                    break;
            }
            if (
                typeof tooltipX === 'number' &&
                typeof tooltipY === 'number' &&
                !isNaN(tooltipY) &&
                !isNaN(tooltipX)
            ) {
                return {
                    left: tooltipX,
                    top: tooltipY,
                    width: isLoaded ? 'auto' : `50px`,
                    height: isLoaded ? 'auto' : `20px`
                };
            }
            return {};
        },
        [
            maxHeight,
            newWidth,
            padding,
            isLoaded,
            hasFailed,
            width,
            height,
            position,
            arrowSize,
            showArrow
        ]
    );

    useEffect(
        () => {
            if (triggerRef && triggerRef.current) {
                setStyles(getTooltipPosition());
            }
        },
        [triggerRef, getTooltipPosition, isHovering]
    );

    const getOppositeSide = useCallback(
        () => {
            switch (position) {
                case 'bottom':
                    return 'top';
                case 'left':
                    return 'right';
                case 'right':
                    return 'left';
                case 'top':
                default:
                    return 'bottom';
            }
        },
        [position]
    );

    const getArrowStyles = useCallback(
        () => {
            switch (position) {
                case 'bottom':
                    return {
                        bottom: '100%',
                        left: '50%',
                        transform: 'translateX(-50%)',
                        borderLeft: `${arrowSize}px solid transparent`,
                        borderRight: `${arrowSize}px solid transparent`,
                        borderBottom: `${arrowSize}px solid rgba(0, 0, 0, 0.8)`
                    };
                case 'right':
                    return {
                        top: '50%',
                        right: '100%',
                        transform: 'translateY(-50%)',
                        borderTop: `${arrowSize}px solid transparent`,
                        borderBottom: `${arrowSize}px solid transparent`,
                        borderRight: `${arrowSize}px solid rgba(0, 0, 0, 0.8)`
                    };
                case 'left':
                    return {
                        top: '50%',
                        left: '100%',
                        transform: 'translateY(-50%)',
                        borderTop: `${arrowSize}px solid transparent`,
                        borderBottom: `${arrowSize}px solid transparent`,
                        borderLeft: `${arrowSize}px solid rgba(0, 0, 0, 0.8)`
                    };
                case 'top':
                default:
                    return {
                        top: '100%',
                        left: '50%',
                        transform: 'translateX(-50%)',
                        borderLeft: `${arrowSize}px solid transparent`,
                        borderRight: `${arrowSize}px solid transparent`,
                        borderTop: `${arrowSize}px solid rgba(0, 0, 0, 0.8)`
                    };
            }
        },
        [position, arrowSize]
    );

    const tooltipElement = (
        <div
            className="image-preview-tooltip"
            style={{
                position: 'absolute',
                zIndex: 9999,
                backgroundColor: 'rgba(0, 0, 0, 0.8)',
                color: '#fff',
                padding: `${padding}px`,
                borderRadius: 3,
                transition: '0.4s ease',
                transitionProperty: 'transform, opacity',
                transformOrigin: `${getOppositeSide(position)} center`,
                transform: 'scale(0)',
                opacity: 0
            }}>
            {isLoaded && !hasFailed ? (
                <img
                    src={src}
                    alt={alt}
                    style={{ width: `${newWidth}px`, height: `${maxHeight}px`, display: 'block' }}
                />
            ) : hasFailed ? null : (
                <LoadingDots
                    style={{
                        marginTop: 3
                    }}
                    size="sm"
                    whiteDots
                />
            )}
            {hasFailed && (
                <span
                    className="-text-center gds-text--body-xs"
                    style={{ width: `${width}px`, height: `${height}px`, display: 'block' }}>
                    Failed to load GIF.
                </span>
            )}
            {showArrow && (
                <span
                    style={{
                        position: 'absolute',
                        width: 0,
                        height: 0,
                        ...getArrowStyles(position)
                    }}
                />
            )}
        </div>
    );
    return (
        <span
            onMouseEnter={() => onEnterLeave(true)}
            onMouseLeave={() => onEnterLeave(false)}
            ref={triggerRef}
            style={{
                position: 'relative',
                cursor: 'pointer',
                display: 'inline-block',
                ...customStyles
            }}>
            {children}
            {ReactDOM.createPortal(
                React.cloneElement(tooltipElement, {
                    style: {
                        ...tooltipElement.props.style,
                        ...styles,
                        transform: isHovering ? 'scale(1)' : 'scale(0)',
                        opacity: isHovering ? 1 : 0
                    }
                }),
                document.body
            )}
        </span>
    );
}
ImagePreview.displayName = 'ImagePreview';
ImagePreview.defaultProps = {
    maxHeight: 200,
    padding: 5,
    position: 'top',
    showArrow: true,
    arrowSize: 5
};
ImagePreview.propTypes = {
    children: PropTypes.node.isRequired,
    src: PropTypes.string.isRequired,
    alt: PropTypes.string,
    maxHeight: PropTypes.number,
    padding: PropTypes.number,
    position: PropTypes.oneOf(['top', 'bottom', 'left', 'right']),
    style: PropTypes.object,
    showArrow: PropTypes.bool,
    arrowSize: PropTypes.number
};

export default ImagePreview;
