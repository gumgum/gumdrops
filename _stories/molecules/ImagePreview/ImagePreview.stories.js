import React from 'react';
import { withKnobs, text, object, number, select, boolean } from '@storybook/addon-knobs';
import mdx from './ImagePreview.mdx';
import Button from '../../../components/atoms/Button';
import ImagePreview from '../../../components/molecules/ImagePreview';

const positionOptions = {
    top: 'top',
    left: 'left',
    right: 'right',
    bottom: 'bottom'
};

export default {
    component: ImagePreview,
    title: 'Molecules/ImagePreview',
    decorators: [withKnobs],
    parameters: {
        component: ImagePreview,
        docs: { page: mdx }
    }
};

export const Default = () => (
    <div
        className="gds-flex gds-flex--justify-center"
        style={{
            padding: 200
        }}>
        <ImagePreview
            className={text('Classname', '')}
            padding={number('Padding', 5)}
            arrowSize={number('Arrow Size', 5)}
            showArrow={boolean('Show Arrow', true)}
            maxHeight={number('Max Height', 200)}
            position={select('Position', positionOptions, 'top')}
            style={object('Style', {})}
            src={text(
                'Src',
                'https://c.gumgum.com/ads/com/gumgum/mantii/internal_mobile_inscreen_test/full_canvas/01/mantii_fullscreen.hyperesources/girl@2x.png'
            )}>
            <Button>Hover Me</Button>
        </ImagePreview>
    </div>
);
