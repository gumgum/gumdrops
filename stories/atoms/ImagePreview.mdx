import { Canvas, Source } from '@storybook/blocks';
import { ArgsTable } from '@storybook/addon-docs';
import * as ImagePreviewStories from './ImagePreview.stories';
import ImagePreview from '../../components/atoms/ImagePreview';

# ImagePreview

The `<ImagePreview>` component is used to preview images or gifs.

It will pre-load the image and show a loading indicator until resolved, then the preview will display.

This can be used to wrap any component and will by default appear at the top of hovering element.


## ImagePreview example
```jsx
<ImagePreview src="PATH_TO_IMAGE">
  <Icon icon="plus" />
</ImagePreview>
```

## Component Example

<Canvas of={ImagePreviewStories.Default} />

## Component Example Source

<Source of={ImagePreviewStories.Default} />

## ImagePreview Props

<ArgsTable of={ImagePreview} />
