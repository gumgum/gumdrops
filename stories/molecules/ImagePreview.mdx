import { Canvas, Source } from '@storybook/blocks';
import { ArgsTable } from '@storybook/addon-docs';
import * as ImagePreviewStories from './ImagePreview.stories';
import ImagePreview from '../../components/molecules/ImagePreview';

# ImagePreview

The `<ImagePreview>` component is used to preview images or gifs.

It will pre-load the image and show a loading indicator until resolved, then the preview will display.


## ImagePreview example
```jsx
<ImagePreview src="PATH_TO_IMAGE">
  <Icon icon="plus" />
</ImagePreview>
```

## Component Example

<Canvas of={ImagePreviewStories.Default} />

## Component Example Source

<Source of={ImagePreviewStories.Default} />

## ImagePreview Props

<ArgsTable of={ImagePreview} />
