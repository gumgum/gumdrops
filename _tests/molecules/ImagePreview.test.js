/* globals mount */
import React from 'react';
import ImagePreview from '../../components/molecules/ImagePreview';

const defaultProps = {
    src:
        'https://c.gumgum.com/ads/com/gumgum/mantii/internal_mobile_inscreen_test/full_canvas/01/mantii_fullscreen.hyperesources/girl@2x.png'
};

describe('Expect <ImagePreview>', () => {
    it('to render', () => {
        const wrapper = mount(
            <div>
                <ImagePreview {...defaultProps}>
                    <input type="text" />
                </ImagePreview>
            </div>
        );
        expect(wrapper).toMatchSnapshot();
    });

    it('should show on hover', () => {
        const wrapper = mount(
            <div>
                <ImagePreview {...defaultProps}>
                    <input type="text" />
                </ImagePreview>
            </div>
        );
        const getTooltip = () => wrapper.find('.image-preview-tooltip');
        const tooltip = getTooltip();
        // tooltip should exist
        expect(tooltip.exists()).toBe(true);
        // expect the opacity to be 0
        expect(tooltip.getDOMNode().style.opacity).toBe('0');
        wrapper.find('ImagePreview > span').simulate('mouseEnter');
        // tooltip should now be visible
        expect(tooltip.getDOMNode().style.opacity).toBe('1');
    });
});
