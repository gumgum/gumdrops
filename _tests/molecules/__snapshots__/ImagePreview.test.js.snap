// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Expect <ImagePreview> to render 1`] = `
<div>
  <ImagePreview
    arrowSize={5}
    maxHeight={200}
    padding={5}
    position="top"
    showArrow={true}
    src="https://c.gumgum.com/ads/com/gumgum/mantii/internal_mobile_inscreen_test/full_canvas/01/mantii_fullscreen.hyperesources/girl@2x.png"
  >
    <span
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      style={
        Object {
          "cursor": "pointer",
          "display": "inline-block",
          "position": "relative",
        }
      }
    >
      <input
        type="text"
      />
      <Portal
        containerInfo={
          <body>
            <div
              class="image-preview-tooltip"
              style="position: absolute; z-index: 9999; background-color: rgba(0, 0, 0, 0.8); color: rgb(255, 255, 255); padding: 5px; border-radius: 3px; transition: 0.4s ease; transition-property: transform, opacity; transform-origin: bottom center; transform: scale(0); opacity: 0; left: -30px; top: -30px; width: 50px; height: 20px;"
            >
              <div
                style="margin-top: 3px;"
              >
                <div
                  class="gds-loading"
                >
                  <div
                    class="gds-loading__dot gds-loading__dot--sm gds-loading__dot--white"
                  />
                </div>
              </div>
              <span
                style="position: absolute; width: 0px; height: 0px; top: 100%; left: 50%; transform: translateX(-50%); border-left: 5px solid transparent; border-right: 5px solid transparent;"
              />
            </div>
          </body>
        }
      >
        <div
          className="image-preview-tooltip"
          style={
            Object {
              "backgroundColor": "rgba(0, 0, 0, 0.8)",
              "borderRadius": 3,
              "color": "#fff",
              "height": "20px",
              "left": -30,
              "opacity": 0,
              "padding": "5px",
              "position": "absolute",
              "top": -30,
              "transform": "scale(0)",
              "transformOrigin": "bottom center",
              "transition": "0.4s ease",
              "transitionProperty": "transform, opacity",
              "width": "50px",
              "zIndex": 9999,
            }
          }
        >
          <LoadingDots
            size="sm"
            style={
              Object {
                "marginTop": 3,
              }
            }
            whiteDots={true}
          >
            <div
              style={
                Object {
                  "marginTop": 3,
                }
              }
            >
              <div
                className="gds-loading"
              >
                <div
                  className="gds-loading__dot gds-loading__dot--sm gds-loading__dot--white"
                />
              </div>
            </div>
          </LoadingDots>
          <span
            style={
              Object {
                "borderLeft": "5px solid transparent",
                "borderRight": "5px solid transparent",
                "borderTop": "5px solid rgba(0, 0, 0, 0.8)",
                "height": 0,
                "left": "50%",
                "position": "absolute",
                "top": "100%",
                "transform": "translateX(-50%)",
                "width": 0,
              }
            }
          />
        </div>
      </Portal>
    </span>
  </ImagePreview>
</div>
`;
